docs/source/topics/scheduler_template.py

```diff
@@ -0,0 +1,182 @@
+# -*- coding: utf-8 -*-
+"""Template for a scheduler plugins."""
+import datetime
+import logging
+
+from aiida.common.escaping import escape_for_bash
+from aiida.schedulers import Scheduler, SchedulerError, SchedulerParsingError
+from aiida.schedulers.datastructures import (JobInfo, JobState, MachineInfo, NodeNumberJobResource, ParEnvJobResource)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+_MAP_SCHEDULER_AIIDA_STATUS = {
+    'SET_ME_STATUS_1': JobState.RUNNING,
+    'SET_ME_STATUS_5': JobState.QUEUED,
+    'SET_ME_STATUS_2': JobState.DONE,
+    'SET_ME_STATUS_3': JobState.QUEUED_HELD,
+    'SET_ME_STATUS_4': JobState.UNDETERMINED,
+    'SET_ME_STATUS_6': JobState.SUSPENDED,
+}
+
+
+class TemplateJobResource(NodeNumberJobResource):  # or TemplateJobResource(ParEnvJobResource)
+    """Template class for job resources."""
+
+    @classmethod
+    def validate_resources(cls, **kwargs):
+        """Validate the resources against the job resource class of this scheduler.
+
+        This extends the base class validator.
+
+        :param kwargs: dictionary of values to define the job resources
+        :return: attribute dictionary with the parsed parameters populated
+        :raises ValueError: if the resources are invalid or incomplete
+        """
+        resources = super().validate_resources(**kwargs)
+
+        # Put some validation code here.
+
+        return resources
+
+
+class TemplateScheduler(Scheduler):
+    """Base class template for a scheduler."""
+
+    # Query only by list of jobs and not by user
+    _features = {
+        'can_query_by_user': False,
+    }
+
+    # The class to be used for the job resource.
+    _job_resource_class = TemplateJobResource
+
+    _map_status = _MAP_SCHEDULER_AIIDA_STATUS
+
+    def _get_resource_lines(
+        self, num_machines, num_mpiprocs_per_machine, num_cores_per_machine, max_memory_kb, max_wallclock_seconds
+    ):
+        """
+        Return a list of lines (possibly empty) with the header
+        lines relative to:
+
+        * num_machines
+        * num_mpiprocs_per_machine
+        * num_cores_per_machine
+        * max_memory_kb
+        * max_wallclock_seconds
+
+        This is done in an external function because it may change in
+        different subclasses.
+        """
+        return []
```

Comment:
This does not actually seem to be required by the base `Scheduler` class, so I would remove it. Did you add for it a particular reason?

Reply:
There was no particular reason. I just tried to make the template as complete as possible. But if you think that it should be removed - it is fine with me.